learning-another-language/js-loops.md

```diff
@@ -0,0 +1,417 @@
+	console.log(arr[i]);
+}
+
+// Output:
+// foo
+// bar
+// baz
+// undefined
+
+// In the above example, arr.length = 3, 
+// so accessing arr[3] leads to an undefined element
+```
+
+When iterating over an array, it is possible to accidentally exceed the bounds of the array. One thing to note about the indices of an array: 
+
+- The first element of an array is stored at index 0. This is true for JavaScript as well as most other programming languages. Therefore, the index of the last element of an array is equal to `arr.length - 1`.
+
+When constructing a `for-loop`, there are a couple ways to handle exceeding the bounds of an array. They both involve altering the conditional statement in the `for-loop` to either:
+
+- `i < arr.length` or
+- `i <= arr.length - 1`
+
+```js
+const arr = [ "foo", "bar", "baz" ];
+
+// Replace the <= with < in the condition statement
+for (let i = 0; i < arr.length; i++) {
+	console.log(arr[i]);
+}
+
+// Output:
+// foo
+// bar
+// baz
+
+// Replace arr.length with arr.length - 1
+for (let i = 0; i <= arr.length - 1; i++) {
+	console.log(arr[i]);
+}
+
+// Output:
+// foo
+// bar
+// baz
+```
+
+## 2.`for … of` loop
+```js
+for (variable of object) {
+	// do something
+}
+```
+
+The `for … of` loop can be used to iterate over arrays, sets, and maps. When using the `for … of` loop, there’s no danger of going out of bounds! The loop will go over each item in the iterable object.
+
+### Examples
+
+**1. Iterate over items in a list**
+```js
+const monsters = [
+	"Michael Myers", 
+	"Jason Voorhees", 
+	"Freddy Krueger"
+];
+
+for (let monster of monsters) {
+	console.log(monster);
+}
+
+// Output:
+// Michael Myers
+// Jason Voorhees
+// Freddy Krueger
+```
+
+**2. Iterate over items in a map**
+```js
+const m = new Map();
+m.set(1, "red");
+m.set(2, "black")
+
+for (let item of m) {
+	console.log(item);
+}
+
+// Output:
+// [1, red]
+// [2, black]
+```
+
+## 3. `for … in` loop
+```js
+for (property in object) {
+	// do stuff
+}
+```
+
+The `for … in` loop iterates over all enumerable properties of an object. The code in the loop is executed for each property in the object. 
+
+### Example
+```js
+const movies = {
+	2008: "The Dark Knight",
+	2009: "Avatar",
+	2010: "Toy Story 3"
+};
+
+for (let key in movies) {
+	console.log(key + ": " + movies[key]);
+}
+
+// Ouput:
+// 2008: The Dark Knight
+// 2009: Avatar
+// 2010: Toy Story 3
+```
+
+### Common Pitfall: Unexpected Behavior When Iterating over an Array
+
+While you can use the `for … in` to iterate over an array, it is recommended to use one of the aforementioned for-loops instead. 
+
+The reason for this is because the `for … in` loop iterates over **all** of the enumerable properties in an object, including any that may be inherited.
+
+If, for instance, a JS library modifies the `Array` prototype, the loop will iterate over any additional properties as well. Check out the example below for a potentially non-intuitive demonstration.
```

Comment:
I'm not sure they are familiar with what a prototype is. We may need to expand on what the prototype property is. It might be enough to say the reasoning has to do with something called prototypes which are out of scope for this course, and include a link so they can follow their curiosity 💫